doc/developer/design/20220404_metadata_storage.md

```diff
@@ -0,0 +1,304 @@
+don't think we need to run PoCs with CockroachDB or FoundationDB. Aurora
+PostgreSQL will scale sufficiently well for several months, if not years; that
+will give us plenty of time down the road to run PoCs with CockroachDB or
+FoundationDB if it becomes necessary.
+
+## Alternatives considered
+
+### Cross-environment reporting queries
+
+The ability to run cross-environment reporting queries has been suggested as
+a hard requirement for the metadata database. There are three known use cases
+for this:
+
+  1. Exposing a global view of the region to the ops team.
+  2. Providing business intelligence via Fivetran-managed CDC to Snowflake.
+  3. Powering reporting jobs like billing.
+
+Essentially, instead of having separate databases for each environment...
+
+```sql
+CREATE DATABASE environment_1;
+CREATE TABLE environment_1.clusters (id int, spec text);
+CREATE TABLE environment_1.view s(id int, sql text);
+
+CREATE DATABASE environment_2;
+CREATE TABLE environment_2.clusters (id int, spec text);
+CREATE TABLE environment_2.views (id int, sql text);
+```
+
+...we'd have a single database where each table has an environment ID:
+
+
+```sql
+CREATE DATABASE aws_us_east_1;
+CREATE TABLE aws_us_east_1.clusters (environment_id int, id int, spec text);
+CREATE TABLE aws_us_east_1.views (environment_id, id int, sql text);
+```
+
+This allows:
+
+  1. The ops team to run queries like `SELECT * FROM clusters` to see all
+     clusters across all customer environments.
+  2. Configuration of a single Fivetran sync job to export all data in this
+     database to Snowflake.
+  3. A single billing job in each region to query the database to determine
+     what clusters were created and deleted across all customer environments.
+
+Unfortunately there are several downsides to this approach:
+
+  * Schema changes would not be isolated to each environment. Every change to the
+    schema would require a two step process: first the new column would need to
+    be added to PostgreSQL; only *then* could environments be upgraded to the
+    new `materialized`. Rollbacks are similarly difficult: it might be
+    impossible to roll back an environment if a backwards-incompatible schema
+    change has been deployed to the region. The logistics of these schema
+    changes would be quite delicate, as the definition of the schema would
+    live in the MaterializeInc/materialize repository but the application of the
+    migration would be managed by MaterializeInc/cloud.
+
+    I believe this point in particular would have deleterious effects on our
+    development velocity, and also seems like the stuff of region-wide outages.
+
+  * We'd need a separate code path to manage migrations in the binary and BYOC
+    deployments.
+
+  * Security would be tricky. We'd need to set up complicated row-based access
+    control policies to limit each environment to accessing only the rows `WHERE
+    environment_id = <allowed_environment_id>`.
```

Comment:
does row-based security cost perf in postgres? might be another downside

Reply:
> One logical database per customer environment provides the most isolation. Makes me the most comfortable from a security perspective.

@benesch should this text be updated to say it's a separate database rather than rows with an `environment_id` column?

Reply:
Ah, this is the section that describes the multi-tenant alternative. I think it is in the correct place?